pkg/mcs/exportspec_test.go

```diff
@@ -78,25 +97,56 @@ func TestCompatibility(t *testing.T) {
 		other      *mcs.ExportSpec
 		compatible bool
 		field      string
+		mutate     func(mcs.ExportSpec) *mcs.ExportSpec
 	}{
-		"compatible: empty": {spec: &mcs.ExportSpec{}, other: &mcs.ExportSpec{}, compatible: true},
-		/*
-			"compatible: identical":        {},
-			"compatible: local properties": {},
-			"conflicting: type":            {},
-			"conflicting: affinity":        {},
-			"conflicting: port":            {},
-		*/
+		"compatible: empty": {spec: &mcs.ExportSpec{}, compatible: true,
+			mutate: func(es mcs.ExportSpec) *mcs.ExportSpec {
+				return &mcs.ExportSpec{}
+			}},
+
+		"compatible: identical": {spec: &sample, compatible: true,
+			mutate: func(es mcs.ExportSpec) *mcs.ExportSpec {
+				return &es
+			}},
+
+		"compatible: local properties": {spec: &sample, compatible: true, //#TODO check if that whay Etai means. Does it
+			mutate: func(es mcs.ExportSpec) *mcs.ExportSpec {
+				es.CreatedAt = metav1.NewTime(time.Now().UTC())
+				es.Namespace = "different_namepsace"
+				es.Name = "different_name"
+				return &es
+			}},
+
+		"conflicting: type": {spec: &sample, compatible: false, field: "type",
+			mutate: func(es mcs.ExportSpec) *mcs.ExportSpec {
+				es.Service.Type = mcsv1a1.Headless
+				return &es
+			}},
+
+		"conflicting: affinity": {spec: &sample, compatible: false, field: "affinity",
+			mutate: func(es mcs.ExportSpec) *mcs.ExportSpec {
+				es.Service.SessionAffinity = corev1.ServiceAffinityNone
+				return &es
+			}},
+
+		"conflicting: port": {spec: &sample, compatible: false, field: "ports",
+			mutate: func(es mcs.ExportSpec) *mcs.ExportSpec {
+				es.Service.Ports = append(es.Service.Ports, mcsv1a1.ServicePort{Port: 81, Name: "http", Protocol: corev1.ProtocolTCP})
+				return &es
+			}},
 	}
 
 	assertions := require.New(t)
 
 	for name, test := range testcases {
 		t.Logf("Running test case %s", name)
 
-		compatible, field := test.spec.IsCompatibleWith(test.other)
-		assertions.Equal(test.compatible, compatible)
-		assertions.Equal(test.field, field, "unexpected field conflict detected")
+		other := test.mutate(*test.spec)
+		errCompatible := test.spec.IsCompatibleWith(other)
+		assertions.Equal(test.compatible, errCompatible == nil) //#TODO make sure its elegant
```

Comment:
I wasn't sure how you wanted me to use the returned error (some casses do need to return the error, so a simple nil check won't do it). I'm not sure if its an elegant way (or clear way) to write it, but I ended up doing it like this

Reply:
Instead of testing compatible and `err == nil`, we can start with `assertions.IsError/IsNil/Error/NotNil` (I don't remember the exact name) and check the return value:
```go
   if test.compatible {
      assertions.NoError(errCompatible)
   else {
      assertions.Error(errCompatible)
   }
```

The exact field name does not matter, but you can add an assertion that the error String is not empty when there's an error. The exact details of the conflict are not interesting from a caller point of view since it'll only set them on the Conflict condition of the `ServiceExport.Status`

Reply:
also, just call `errCompatible`  - `err`. No need to be elaborate as it is the only error in the test and you check it immediately after calling.

Reply:
> The exact field name does not matter, but you can add an assertion that the error String is not empty when there's an error. The exact details of the conflict are not interesting from a caller point of view since it'll only set them on the Conflict condition of the `ServiceExport.Status`

Shouldn't we make sure that the field that we match as error is the right one (and not accidentally raised error on a different field)?

Reply:
I'm debating this with myself... 
From caller perspective it does not really matter, but I agree that detecting the right field should be validated.
How about defining (string) constants for the fields (e.g., NameField, PortField, ...) that can conflict, adding a `Cause()` or `Field()` to the CompatibilityError class so we can compare against that?